text/0000-glimmer-component-signature.md

```diff
@@ -0,0 +1,361 @@
+
+[official TypeScript support]: https://github.com/emberjs/rfcs/pull/724
+[type-aware rules]: https://github.com/typescript-eslint/typescript-eslint#can-we-write-rules-which-leverage-type-information
+
+## Drawbacks
+
+As with any change that deprecates supported behavior, there's an inherent cost associated with migrating the user base over to new patterns. One goal of this change, however, is to ease such migrations in the future: as the templating system evolves and new information becomes relevant to capture and document, the `Signature` type provides a place for that information to live without disrupting existing code.
+
+The other potential drawback to this approach is that it introduces TypeScript type information that has no visible effect on the component using out-of-the-box tooling. Without a template-aware system like Glint or `els-addon-typed-templates` for validating components, nothing enforces that the signature declared is actually accurate. See the "Only `Args`" section under Alternatives below for further discussion of this point.
+
+## Alternatives
+
+### Additional Positional Type Parameters
+
+Rather than wrapping the information we're interested in capturing in a `Signature` type, we could instead introduce further type parameters to the `Component` base class:
+
+```ts
+class Component<Args = {}, BlockParams = {}, Element = null> {
+  // ...
+}
+```
+
+This has the advantage of not requiring current users of the `Args` parameter to change their code, but suffers many of the same ergonomic issues as regular functions do when they begin to accrue many positional parameters. Authors need to remember which parameters appear in which order when using the `Component` type, and they need to be aware of the appropriate default values for earlier parameters to fill in when they only want to specify a later one.
+
+
+### Only `Args`
+
+One of the drawbacks mentioned above is that the `Element` and `BlockParams` signature members described in this RFC are functionally inert in a vanilla TypeScript project. This leaves them with about the same status as comments: potentially helpful when left by a well-meaning author, but without any checks to ensure that they're accurate and that they stay up-to-date as the implementation changes.
+
+An alternative would be to still introduce the `Signature` type in `@glimmer/component` but _only_ formalize the `Args` member, leaving other tooling to define the semantics of any additional signature members they might be interested in. While this would simplify the overall proposal somewhat, what a component `{{yield}}`s and what it does with its `...attributes` are core enough to a component's public interface that we believe they should be considered first-class rather than having individual tools reinvent them, potentially in mutually-incompatible ways.
```

Comment:
`## Unresolved Questions`

Would add the `Attributes` here (copy pasting the paragraph from my draft) for _completeness_ and a clear statements they are out of scope for now but we provide the option to later include them.

Reply:
This document discusses in a couple different spots the fact that we leave the door open to future additions to `Signature`, but the `Attributes` idea itself _isn't_ an unresolved question of this RFC. I don't think discussing `Attributes`, even just to explicitly state that it's out of scope, fills in any holes in the case being made in this document.

Best case, I don't think it adds anything, and worst case (given how divisive the idea is, both in terms of what it should capture and whether it's a good addition at all) it sparks its own discussion that distracts from the actual proposal in the RFC.